### PR TITLE
Generate: replace breaks by a loop condition

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Pulled from the `torch.compile(..., fullgraph=True)` draft PR: https://github.com/huggingface/transformers/pull/29374

It replaces the `breaks` that exit the endless generation loop with an equivalent function that returns `False` when it should stop generating, while preserving ZeRO stage 3 support. It is not only an improvement in terms of code reuse, but also a hard requirement to enable  `torch.compile(..., fullgraph=True)`: `break` and data-dependent control flow is not supported.

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -2187,12 +2196,7 @@ def _contrastive_search(
 
             # stop when each sentence is finished
             unfinished_sequences = unfinished_sequences & ~stopping_criteria(input_ids, scores)
-
-            if unfinished_sequences.max() == 0:
-                this_peer_finished = True
-
-            if this_peer_finished and not synced_gpus:
-                break
+            this_peer_finished = unfinished_sequences.max() == 0
```
Comment (gante) [inline src/transformers/generation/utils.py:2199]: The previous version is also a data-dependent control flow, so this change is for `torch.compile` readiness :)

**Comment by gante** [comment]
FYI @zucchini-nlp

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1778,6 +1778,25 @@ def typeerror():
...
+            # The following logic allows an early break if all peers finished generating their sequence
+            this_peer_finished_flag = torch.tensor(0.0 if this_peer_finished else 1.0).to(device)
+            # send 0.0 if we finished, 1.0 otherwise
+            dist.all_reduce(this_peer_finished_flag, op=dist.ReduceOp.SUM)
+            # did all peers finish? the reduced sum will be 0.0 then
+            if this_peer_finished_flag.item() == 0.0:
+                return False
+        else:
+            if this_peer_finished:
+                return False
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:1797]: nit

```suggestion
        elif this_peer_finished:
            return False
```

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -1778,6 +1778,25 @@ def typeerror():
...
+        return True
```
Comment (amyeroberts) [inline src/transformers/generation/utils.py:1798]: Or actually, we can just do 

```suggestion
        return not this_peer_finished
```

**Review by amyeroberts — APPROVED** [review]
Thanks for working on this - so much cleaner 🤩

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/utils.py:1798]: (we have to go with the other suggestion: in this one, we can return `False` when `synced_gpus=True` and `this_peer_finished` has different values in different GPUs. In this case, it should return `True` to continue generating until all devices are done.)

File: src/transformers/generation/utils.py (hunk 2 above)
Comment (gante) [reply src/transformers/generation/utils.py:1797]: no good: if `synced_gpus` is `True`, we only want to return `False` when `this_peer_finished is True` in all distributed devices. With this change, we can return `False` when `this_peer_finished is True` in a single device :)

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/utils.py:1798]: (see comment above)

Comment (gante) [reply src/transformers/generation/utils.py:1798]: This solution can return `False` when `synced_gpus is True` and `this_peer_finished is True`, which is not intended -- `this_peer_finished` has to be `True` in all distributed devices when `synced_gpus is True` 🤗

**Pull request merged**